From 0f54d7c2f421257c7b1c39d1c39472e40a27980b Mon Sep 17 00:00:00 2001
From: Lauri Alanen <laurialanen1@gmail.com>
Date: Wed, 14 Oct 2026 20:04:20 +0300
Subject: [PATCH] Logger: provide fmt::ostream_formatter specializations for
 fmt >= 9

fmt 9+ removed implicit std::ostream operator<< formatting and requires
explicit fmt::formatter<T> specializations inheriting from
fmt::ostream_formatter.

Upstream-Status: Inappropriate [upstream superbuild pins spdlog 1.9.2 / fmt 8]
---
 include/uxr/agent/logger/Logger.hpp | 40 +++++++++++++++++++++++++++++
 1 file changed, 40 insertions(+)

diff --git a/include/uxr/agent/logger/Logger.hpp b/include/uxr/agent/logger/Logger.hpp
index fe4508cc..b6bbe114 100644
--- a/include/uxr/agent/logger/Logger.hpp
+++ b/include/uxr/agent/logger/Logger.hpp
@@ -24,6 +24,46 @@
 #include <spdlog/fmt/ostr.h>
 #include <spdlog/fmt/bin_to_hex.h>
 #include <spdlog/sinks/stdout_sinks.h>
+
+// fmt >= 9 no longer formats types through operator<< implicitly; opt in
+// for the types the agent logs.
+#if defined(FMT_VERSION) && FMT_VERSION >= 90000
+namespace eprosima {
+namespace uxr {
+class IPv4EndPoint;
+class IPv6EndPoint;
+class SerialEndPoint;
+class MultiSerialEndPoint;
+class CanEndPoint;
+class CustomEndPoint;
+} // namespace uxr
+} // namespace eprosima
+namespace eprosima {
+namespace fastrtps {
+namespace rtps {
+struct EntityId_t;
+struct GuidPrefix_t;
+class SampleIdentity;
+} // namespace rtps
+} // namespace fastrtps
+} // namespace eprosima
+namespace dds {
+namespace xrce {
+class TransportAddress;
+} // namespace xrce
+} // namespace dds
+
+template <> struct fmt::formatter<eprosima::uxr::IPv4EndPoint> : fmt::ostream_formatter {};
+template <> struct fmt::formatter<eprosima::uxr::IPv6EndPoint> : fmt::ostream_formatter {};
+template <> struct fmt::formatter<eprosima::uxr::SerialEndPoint> : fmt::ostream_formatter {};
+template <> struct fmt::formatter<eprosima::uxr::MultiSerialEndPoint> : fmt::ostream_formatter {};
+template <> struct fmt::formatter<eprosima::uxr::CanEndPoint> : fmt::ostream_formatter {};
+template <> struct fmt::formatter<eprosima::uxr::CustomEndPoint> : fmt::ostream_formatter {};
+template <> struct fmt::formatter<dds::xrce::TransportAddress> : fmt::ostream_formatter {};
+template <> struct fmt::formatter<eprosima::fastrtps::rtps::EntityId_t> : fmt::ostream_formatter {};
+template <> struct fmt::formatter<eprosima::fastrtps::rtps::GuidPrefix_t> : fmt::ostream_formatter {};
+template <> struct fmt::formatter<eprosima::fastrtps::rtps::SampleIdentity> : fmt::ostream_formatter {};
+#endif
 #endif
 
 #ifdef _WIN32
-- 
2.43.0

